Sprint 06

Heamex opened the pull request.

Heamex committed:
Прошел урок, подготовил проект к работе с сетью, добавлена возможность показывать Alert если при ошибке загрузки из сети.

Heamex committed:
Завершен спринт 6. Приложение теперь корректно подтягивает данные с сервера IMDB, в случае проблем с загрузкой данных о фильмах выдаёт алерт с ошибкой сетевого подключения.

helecta commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -51,6 +63,17 @@ final class MovieQuizViewController: UIViewController, QuestionFactoryDelegate,
 	
+	// Запускаем индикатор загрузки
+	private func showLoadingIndicator() {
+		activityIndicator.isHidden = false
```
Рекомендую посмотреть на свойство [hidesWhenStopped](https://developer.apple.com/documentation/uikit/uiactivityindicatorview/1622837-hideswhenstopped) 
у activityIndicator, и использовать его вместо ручного управления свойством isHidden. 
Когда ты будешь вызывать startAnimating – индикатор будет появляться автоматически, соответственно на stopAnimating – скрываться, а свойство isHidden больше не понадобится, индикатор все сделает за тебя. 
Эту настройку делают всего раз раз: кодом во viewDidLoad или галочкой у элемента на сториборде.

Heamex replied:
Годный совет. Сделал!

helecta commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -98,6 +121,16 @@ final class MovieQuizViewController: UIViewController, QuestionFactoryDelegate,
+										 text: message,
+										 buttonText: "Попробовать ещё раз")
+		
+		alertPresenter?.showAlert(model: model)
```
Если произошла ошибка загрузки данных, появится алерт. После нажатия на кнопку алерта, будет вызван метод делегата(т.е контроллера) `didAlertButtonPressed`. В нем происходит запрос следующего вопроса `requestNextQuestion`. Но т.к данные не загрузились, то ничего не произойдет. Тут надо, что бы еще раз пробовали загрузить данные.

helecta commented on Services/NetworkClient.swift:
```diff
@@ -0,0 +1,69 @@
+import UIKit
+
+struct MostPopularMovies: Codable {
+	let errorMessage: String
```
Обращу внимание на свойство `errorMessage`. 

Здесь может прийти текст с ошибкой, в таком случае сервис обычно возвращает пустой список фильмов `items`. 
С пустым списком приложение будет выглядеть сломанным. 
Это состояние называют краевым случаем, его стоит обрабатывать как ошибку, например, показывать алерт.
С точки зрения проектирования API сервиса IMDB, 
эту ошибку лучше вернуть c кодом ошибки клиента ([группа 4хх](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status#client_error_responses)). 
Но случается и так, что ответ имеет статус код 200, а тело ответа содержит сообщение об ошибке, как в нашем примере. 
Такой ответ в любом случае необходимо обработать разработчику фронтенда (web, iOS, android).

 **Примеры состояний, когда такое происходит:**

- API key неверный
- API key просрочен
- количество запросов в день превышено (по документации не больше 100)

helecta commented on Services/QuestionFactory.swift:
```diff
@@ -8,32 +8,72 @@
-		QuizQuestion(image: "The Ice Age Adventures of Buck Wild", text: "Рейтинг этого фильма больше чем 6?", correctAnswer: false),
-		QuizQuestion(image: "Tesla", text: "Рейтинг этого фильма больше чем 6?", correctAnswer: false),
-		QuizQuestion(image: "Vivarium", text: "Рейтинг этого фильма больше чем 6?", correctAnswer: false)]
+	/*
```
Закомментированный код лучше удалять, т.к это улучшает читаемость кода.

helecta commented on Services/QuestionFactory.swift:
```diff
@@ -8,32 +8,72 @@
+			do {
+				imageData = try Data(contentsOf: movie.resizedImageURL)
+			} catch {
+				print("Failed to load image")
```
Обращу внимание, что игра состоит из вопроса о рейтинге, на который отвечает пользователь, и контекста о фильме в виде постера(картинка). 
Без постера, пользователю будет затруднительно ответить на вопрос о рейтинге фильма верно.

В этом месте, функция не завершает свою работу и идет формировать вопрос без постера. 
В таком случае, бизнес-логика приложения будет нарушена, потому что пользователь не сможет ответить корректно, ведь изображения нет.
Это состояние стоит обработать, например, с помощью алерта. 
Не забудь, что алерт будет обновлять UI и это тоже нужно учесть (вернуть на main), ровно также, как и прерывание функции (сделать return)

Heamex committed:
Поправил

Добавлен функционал показа ошибки, в случае, если интернет пропал между вопросами, а также добавлена генерация вопроса из трёх вариантов.

Heamex committed:
Update Main.storyboard

Heamex committed:
первые наработки по 7 спринту
